src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1757,35 +1786,43 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
 
 int TheoryStringsRewriter::componentContains(std::vector<Node>& n1,
                                              std::vector<Node>& n2,
-                                             std::vector<Node>& nr,
-                                             bool computeRemainder)
+                                             std::vector<Node>& nb,
+                                             std::vector<Node>& ne,
+                                             bool computeRemainder,
+                                             int remainderDir)
 {
-  if (n2.size() == 1 && n2[0].isConst())
+  // if n2 is a singleton, we can do optimized version here
+  if (n2.size() == 1)
   {
-    CVC4::String t = n2[0].getConst<String>();
     for (unsigned i = 0; i < n1.size(); i++)
     {
-      if (n1[i].isConst())
+      Node n1rb;
+      Node n1re;
+      if (componentContainsBase(n1[i], n2[0], n1rb, n1re, 0, computeRemainder))
       {
-        CVC4::String s = n1[i].getConst<String>();
-        size_t f = s.find(t);
-        if (f != std::string::npos)
+        if (computeRemainder)
         {
-          if (computeRemainder)
+          if (remainderDir != -1)
           {
-            Assert(s.size() >= f + t.size());
-            if (s.size() > f + t.size())
+            if (!n1re.isNull())
             {
-              nr.push_back(NodeManager::currentNM()->mkConst(::CVC4::String(
-                  s.substr(f + t.size(), s.size() - (f + t.size())))));
+              ne.push_back(n1re);
             }
-            nr.insert(nr.end(), n1.begin() + i + 1, n1.end());
-            n1[i] = NodeManager::currentNM()->mkConst(
-                ::CVC4::String(s.substr(0, f + t.size())));
+            ne.insert(ne.end(), n1.begin() + i + 1, n1.end());
             n1.erase(n1.begin() + i + 1, n1.end());
           }
-          return i;
+          n1[i] = n2[0];
```

Comment:
This seems wrong. The invariant is of this function is according to its comment that:

```
n1 = str.++( nb, n1', ne )
```

So we should set `n1[i]` to the opposite of what we put in `ne`/`nb` (i.e. the opposite of `n1re` and `n1rb`, depending on remainderDir).

Reply:
Actually I think this code is correct.  The opposite of what we put in ne / nb is in fact n2[0].

Reply:
My understanding is that `n1[i] = str.++(n1rb, n2[0], n1re)`, so in the case where `n1rb` is non-empty, we should also add that to `n1[i]`, not only `n2[0]` (for `remainderDir = 1`).

Reply:
Ah, I see what you mean, yes, for remainderDIr != 0, it needs to keep one or the other, I'll update

Reply:
good catch btw.  I've fixed this in my latest commit.